data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/impl/auth/OIDCDiscoveryConfigListener.java

```diff
@@ -62,14 +62,10 @@ public OIDCDiscoveryConfigListener(String featuresConfigPath, Vertx vertx, int t
     }
 
     public OIDCDiscoveryConfig getOidcDiscoveryConfig() {
-        return oidcDiscoveryConfig;
+        return oidcDiscoveryConfig.get();
     }
 
-    public int registerCallback(Consumer<OIDCDiscoveryConfig> callback) {
-        if (this.callbacks == null) {
-            this.callbacks = new ArrayList<>();
-        }
-
+    public synchronized int registerCallback(Consumer<OIDCDiscoveryConfig> callback) {
```

Comment:
One other issue I see with index based registration is that with addition and removal of callbacks the array is ever growing which can lead to OOM, it's better to have a map `id -> string`

Reply:
you mean like `ConcurrentMap` ? instead of the `CopyOnWriteArrayList`?

Reply:
yes